### PR TITLE
Implemented Profile, Analyze and Report Subcommands.

### PR DESCRIPTION
**Pull request by aramesh7**

### CODE REVIEW

File: model_analyzer/state/analyzer_state_manager.py [hunk 1]
```diff
@@ -150,8 +150,14 @@ def interrupt_handler(self, signal, frame):
...
-            'Received SIGINT. Stopping profiling and proceeding to process results.'
-        )
+            f'Received SIGINT {self._exiting}/{MAX_NUMBER_OF_INTERRUPTS}. '
+            'Will attempt to exit after current measurement.')
+        if self._exiting >= MAX_NUMBER_OF_INTERRUPTS:
+            logging.info(
+                f'Received SIGINT maximum number of times. Saving state and exiting immediately. '
+                'perf_analyzer and tritonserver may still be running...')
+            self.save_checkpoint()
+            sys.exit(0)
```
Comment (Tabrizian) [inline model_analyzer/state/analyzer_state_manager.py:160]: sys.exit(1)

File: qa/L0_results/check_results.py [hunk 2]
```diff
@@ -18,67 +18,94 @@
...
-                        '--config-file',
-                        type=str,
-                        required=True,
-                        help='The config file for this test')
-    parser.add_argument('-d',
-                        '--export-path',
-                        type=str,
-                        required=True,
-                        help='The full path to the export directory')
+class TestOutputValidator:
```
Comment (Tabrizian) [inline qa/L0_results/check_results.py:21]: Can't comment on the file itself. I think  qa/L0_results/1.ckpt  was mistakenly added.

**Review by Tabrizian — CHANGES_REQUESTED** [review]

File: model_analyzer/config/input/config_command_analyze.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+                    schema={
+                        'objectives': objectives_scheme,
+                        'constraints': constraints_scheme,
+                    })
+            },
+            output_mapper=ConfigModelAnalysisSpec.
+            model_object_to_config_model_analysis_spec)
+        self._add_config(
+            ConfigField(
+                'analysis_models',
```
[the remark this answers is not in the document]
Comment (aramesh7) [reply model_analyzer/config/input/config_command_analyze.py:147]: I wanted to differentiate this field across the various model configs. So we have `profile_models` `analysis_models` and `report_model_configs`

File: model_analyzer/config/input/config_command_analyze.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+                description='Specifies column keys for model gpu metrics table'
+            ))
+        self._add_config(
+            ConfigField('server_output_fields',
+                        field_type=ConfigListString(),
+                        default_value=DEFAULT_SERVER_OUTPUT_FIELDS,
+                        description=
+                        'Specifies column keys for server-only metrics table'))
+
+    def _add_report_configs(self):
```
[the remark this answers is not in the document]
Comment (aramesh7) [reply model_analyzer/config/input/config_command_analyze.py:233]: Thank you. The original infrastructure was what made it easy!

File: model_analyzer/reports/report_manager.py [hunk 5]
```diff
@@ -76,78 +84,206 @@ def data(self, report_key):
...
+                                     'latency_breakdown.png')
+        detailed_caption = f"Latency Breakdown for Online Performance of {report_key}"
+
+        # First add row of detailed
+        detailed_report.add_images([detailed_plot], [detailed_caption])
+
+        # Next add the simple plots created for this detailed report
+        plot_stack = []
+        caption_stack = []
+        plot_path = os.path.join(self._config.export_path, 'plots', 'simple',
```
[the remark this answers is not in the document]
Comment (aramesh7) [reply model_analyzer/reports/report_manager.py:235]: Two parameters x vs y. It is one of the types of plots we can make. Should i update the comment to be more descriptive?

File: model_analyzer/config/input/config_command_analyze.py [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+                f"--export-path not specified. Using {self._fields['export_path'].default_value()}"
+            )
+        elif self.export_path and not os.path.isdir(self.export_path):
+            raise TritonModelAnalyzerException(
+                f"Export path {self.export_path} is not a directory.")
+
+        if self.num_top_model_configs > 0 and not self.constraints:
+            raise TritonModelAnalyzerException(
+                "If setting num_top_model_configs > 0, comparison across models is requested. "
+                "This requires that global constraints be specified in the config to be used as default."
```
[the remark this answers is not in the document]
Comment (aramesh7) [reply model_analyzer/config/input/config_command_analyze.py:285]: The constraints are applied in the same phase as the sorting. Its a good question as to whether we should require global constraints if comparing across models. what do you think?

File: model_analyzer/config/input/config_command_analyze.py [hunk 7]
```diff
@@ -0,0 +1,364 @@
+# Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
```
[the remark this answers is not in the document]
Comment (aramesh7) [reply model_analyzer/config/input/config_command_analyze.py:1]: I wanted the filenames to match the class names, and I didn't want these classes to be called `Analyze`, `Profile` and `Report`, as it was too misleading as to the class purpose.

File: model_analyzer/config/input/config_command_report.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+import logging
+import os
+
+
+class ConfigCommandReport(ConfigCommand):
+    """
+    Model Analyzer config object.
+    """
+    def __init__(self):
```
[the remark this answers is not in the document]
Comment (aramesh7) [reply model_analyzer/config/input/config_command_report.py:39]: Removed

File: model_analyzer/plots/plot_manager.py [hunk 9]
```diff
@@ -12,85 +12,175 @@
...
+        for plot_config in self._config.plots:
+            constraints = self._constraints['default']
+            if plots_key in self._constraints:
+                constraints = self._constraints[plots_key]
+            for result in self._result_manager.top_n_results(
+                    model_name=model_name, n=num_results):
+                if result.model_config().cpu_only():
+                    if plot_config.y_axis() == 'gpu_used_memory':
+                        plot_name, plot_config_dict = list(
+                            DEFAULT_CPU_MEM_PLOT.items())[0]
```
[the remark this answers is not in the document]
Comment (aramesh7) [reply model_analyzer/plots/plot_manager.py:96]: Because the plots for the `analyze` subcommand are not configurable by the user and are specified on a global basis. Making it per model, but not configurable was difficult.

File: model_analyzer/plots/plot_manager.py [hunk 10]
```diff
@@ -12,85 +12,175 @@
...
+        Constructs detailed plots based on
+        requested config specs
+        """
+
+        # Create detailed plots
+        for model in self._config.report_model_configs:
+            model_config_name = model.model_config_name()
+            self._detailed_plots[model_config_name] = DetailedPlot(
+                f'latency_breakdown', 'Online Performance')
+            measurements = self._result_manager.get_model_config_measurements(
```
[the remark this answers is not in the document]
Comment (aramesh7) [reply model_analyzer/plots/plot_manager.py:142]: Very good catch

File: model_analyzer/record/metrics_manager.py [hunk 11]
```diff
@@ -92,10 +93,12 @@ def profile_server(self):
         TritonModelAnalyzerException
         """
 
-        self._start_monitors()
-        if not self._is_cpu_only:
+        cpu_only = (not numba.cuda.is_available())
```
[the remark this answers is not in the document]
Comment (aramesh7) [reply model_analyzer/record/metrics_manager.py:96]: Profiling the server must rely on numba because it is system specific / GPU specific and not model specific.

File: model_analyzer/reports/report_manager.py [hunk 12]
```diff
@@ -76,78 +84,206 @@ def data(self, report_key):
...
+                                 report_key)
+        for plot_config in report_model_config.plots():
+            plot_stack.append(
+                os.path.join(plot_path, f"{plot_config.name()}.png"))
+            caption_stack.append(
+                f"{plot_config.title()} curves for config {report_key}")
+            if len(plot_stack) == 2:
```
[the remark this answers is not in the document]
Comment (aramesh7) [reply model_analyzer/reports/report_manager.py:242]: Yes, its just to guarantee two on each image row. there is one outside the loop for odd numbers of plots.

File: model_analyzer/reports/report_manager.py [hunk 13]
```diff
@@ -309,9 +511,13 @@ def _get_gpu_stats(self, report_key):
         gpu_names = []
         max_memories = []
         seen_gpus = set()
+        cpu_only = True
 
-        for _, measurement in self._data[report_key]:
+        for model_config, measurement in self._summary_data[report_key]:
+            if model_config.cpu_only():
+                continue
             for gpu in cuda.gpus:
```
[the remark this answers is not in the document]
Comment (aramesh7) [reply model_analyzer/reports/report_manager.py:519]: Ahhh this is tough to get around. I think we should handle this in another PR. Filed a [ticket](https://jirasw.nvidia.com/browse/DLIS-2343)

File: qa/L0_results/check_results.py (hunk 2 above)
Comment (aramesh7) [reply qa/L0_results/check_results.py:21]: This was not by accident. Some of the tests now contain checkpoints for speed. They do not need to profile, as they test only on `analyze` or `report` subcommands.

File: model_analyzer/config/input/config_command_analyze.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Tabrizian) [reply model_analyzer/config/input/config_command_analyze.py:285]: I think it shouldn't be required by I guess we don't have time for it.

File: model_analyzer/record/metrics_manager.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Tabrizian) [reply model_analyzer/record/metrics_manager.py:96]: I see.

File: qa/L0_output_fields/test.sh [hunk 14]
```diff
@@ -99,14 +99,14 @@ for CONFIG_FILE in ${LIST_OF_CONFIG_FILES[@]}; do
             RET=1
         fi
         
-        check_log_table_row_column $ANALYZER_LOG $SERVER_METRICS_NUM_COLUMNS ${#GPUS[@]} "Server\ Only:"
```
Comment (Tabrizian) [inline qa/L0_output_fields/test.sh]: Why did you remove `#GPUS[@]`? The test will fail if it lands on a multi gpu machine.

Comment (aramesh7) [reply qa/L0_output_fields/test.sh]: It is not supposed to do a profile for this test. This test only tests the analyze portion, so it loads data from a checkpoint. Th checkpoint only has 1 measurement, for which we are adjusting the output fields

**Review by Tabrizian — APPROVED** [review]

**Review by Tabrizian — APPROVED** [review]
